### PR TITLE
[SW] Enable Additional Compiler Warnings

### PR DESCRIPTION
**Pull request by lenary**
This PR updates the enabled C/C++ compiler warnings to match our coding style guide as much as possible. 

This effort has caught at least one bug (in `dif_i2c.c`), and caught a bunch of places where we were (unknowingly) using GNU extensions. 

There are four warnings we currently turn off, as we currently violate them and they are not trivial to solve. I intend that we do solve them, in the future.

### CODE REVIEW

**Comment by gkelly** [comment]
`cc: error: unrecognized command line option ‘-Wimplicit-fallthrough’`
Is this because this is using gcc 5.5.0?

**Comment by lenary** [comment]
Yeah, I'm working on the meson fix (which won't be pretty).

**Comment by imphil** [comment]
Expect GCC 7.4 coming up hopefully shortly in #3664.

**Comment by imphil** [comment]
Together with the ability to use the latest meson version, if that helps.

**Review by gkelly — APPROVED** [review]
The meson mangling was less intense than I feared! Thanks for figuring this out and getting the extra warnings enabled, @lenary!

File: sw/device/tests/aes_test.c [hunk 1]
```diff
@@ -33,7 +33,7 @@ static const uint8_t kKeyShare1[32] = {
     0xbf, 0xcf, 0xdf, 0xef, 0xff, 0x0a, 0x1a, 0x2a, 0x3a, 0x4a, 0x5a,
     0x6a, 0x7a, 0x8a, 0x9a, 0xaa, 0xba, 0xca, 0xda, 0xea, 0xfa};
 
-const test_config_t kTestConfig = {};
+const test_config_t kTestConfig;
```
Comment (mcy) [inline sw/device/tests/aes_test.c:36]: Oh, is writing `= {}` instead of `= {0}` a GNU extension? TIL!

**Review by mcy — APPROVED** [review]
This is great, thanks Sam.

File: sw/device/tests/aes_test.c (hunk 1 above)
Comment (lenary) [reply sw/device/tests/aes_test.c:36]: Yeah, this comes from the C grammar not actually allowing initialiser lists to be empty.

**Pull request merged**